### PR TITLE
fix(rust): Untangle offset commit on revocation

### PR DESCRIPTION
**Pull request by loewenheim**
This changes the `AssignmentCallback` trait so that the `on_revoke` method takes an additional generic parameter `commit_offsets` that takes care of committing offsets before the revocation. This removes a circularity in the `StreamProcessor`.

Unfortunately this requires us to do some more unholy things in the `KafkaConsumer`: `KafkaConsumer` contains a `BaseConsumer`, which in turn contains a `CustomContext`. This PR changes it so that the `CustomContext` _also_ has an `Arc<Mutex>` to the `BaseConsumer`, creating a reference cycle. This is because there is no real other way to pass the `BaseConsumer` (or any `rdkafka` client) to the `CustomContext`.

As a sort-of drive by, this also removes some dynamic dispatch. For example, `Consumer` now has another type parameter for the callback type.

### CODE REVIEW

**Review by untitaker — APPROVED** [review]

**Review by Swatinem — APPROVED** [review]
this is a bit hard to follow, but looks okay. I’m a bit surprised Rust allows having a generic inside of a trait method, and I also don’t quite understand why we would need that?

**Comment by loewenheim** [comment]
> this is a bit hard to follow, but looks okay. I’m a bit surprised Rust allows having a generic inside of a trait method, and I also don’t quite understand why we would need that?

Having a generic parameter in a trait method isn't a problem, the trait just won't be object safe.
